presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,449 @@
+            QualifiedName.of("MIN"),
+            QualifiedName.of("MAX"),
+            QualifiedName.of("SUM"));
+
+    private MaterializedViewInfo materializedViewInfo;
+    private Optional<Identifier> removablePrefix = Optional.empty();
+    private Optional<Set<Expression>> expressionsInGroupBy = Optional.empty();
+
+    public MaterializedViewQueryOptimizer(
+            Metadata metadata,
+            Session session,
+            SqlParser sqlParser,
+            AccessControl accessControl,
+            RowExpressionDomainTranslator domainTranslator,
+            Table materializedView,
+            Query materializedViewQuery)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.session = requireNonNull(session, "session is null");
+        this.sqlParser = requireNonNull(sqlParser, "sql parser is null");
+        this.accessControl = requireNonNull(accessControl, "access control is null");
+        this.domainTranslator = requireNonNull(domainTranslator, "row expression domain translator is null");
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+        FunctionAndTypeManager functionAndTypeManager = metadata.getFunctionAndTypeManager();
+        logicalRowExpressions = new LogicalRowExpressions(
+                new RowExpressionDeterminismEvaluator(functionAndTypeManager),
+                new FunctionResolution(functionAndTypeManager),
+                functionAndTypeManager);
+    }
+
+    public Node rewrite(Node node)
+    {
+        // TODO: Implement ways to handle non-optimizable query without throw/catch. https://github.com/prestodb/presto/issues/16541
+        try {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.warn(ex.getMessage());
+            return node;
+        }
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        for (Node child : node.getChildren()) {
+            process(child, context);
+        }
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        return new Query(
+                node.getWith(),
+                (QueryBody) process(node.getQueryBody(), context),
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 242-250 into a separate method.